### PR TITLE
peer: include verack as part of version handshake

### PR DESCRIPTION
**Pull request by wpaulino**
It was possible for connections to bitcoind nodes to be closed from their side if the OnVersion callback queued a message to send. This is because bitcoind expects a verack message before attempting to process any other messages. To address this, we ensure the verack is sent as part of the handshake process, such that any queued messages happen after the fact.

### CODE REVIEW

File: peer/peer.go [hunk 1]
```diff
@@ -2043,29 +2069,60 @@ func (p *Peer) writeLocalVersionMsg() error {
 		return err
 	}
 
-	return p.writeMessage(localVerMsg, wire.LatestEncoding)
+	if err := p.writeMessage(localVerMsg, wire.LatestEncoding); err != nil {
```
Comment (Roasbeef) [inline peer/peer.go]: Unnecessary change.

File: peer/peer.go [hunk 2]
```diff
@@ -2043,29 +2069,60 @@ func (p *Peer) writeLocalVersionMsg() error {
...
-// then sends our version message. If the events do not occur in that order then
-// it returns an error.
+// negotiateInboundProtocol performs the negotiation protocol for an inbound
+// peer. The events should occur in the following order, otherwise an error is
+// returned:
+//
+//   1. Remote peer sends their version.
+//   2. We send our version.
+//   3. We send our verack.
+//   4. Remote peer sends their verack.
```
Comment (Roasbeef) [inline peer/peer.go]: We may want to add some grace to this ordering here, to ensure that both sides don't deadlock if we're connecting to an implementation with a slightly different ordering.

File: peer/peer.go [hunk 3]
```diff
@@ -1974,6 +1966,40 @@ func (p *Peer) readRemoteVersionMsg() error {
...
+
+	// It should be a verack message, otherwise send a reject message to the
+	// peer explaining why.
+	msg, ok := remoteMsg.(*wire.MsgVerAck)
+	if !ok {
+		reason := "a verack message must follow version"
+		rejectMsg := wire.NewMsgReject(
+			msg.Command(), wire.RejectMalformed, reason,
+		)
+		_ = p.writeMessage(rejectMsg, wire.LatestEncoding)
```
Comment (Roasbeef) [inline peer/peer.go]: We may want to log this error at the very least.

Comment (wpaulino) [reply peer/peer.go]: The error will propagate back to `peer.AssociateConnection` which will log it.

File: peer/peer.go (hunk 2 above)
Comment (wpaulino) [reply peer/peer.go]: If they send a message in the wrong order, then an error would be returned which would cause us to disconnect them. If they don't sent a message we expect, there's a timeout that'll also cause us to disconnect them. I believe that should cover everything?

File: peer/peer.go (hunk 1 above)
Comment (wpaulino) [reply peer/peer.go]: Fixed.

File: peer/peer.go (hunk 2 above)
Comment (Roasbeef) [reply peer/peer.go]: Yeah it's also already the case that we have an expected order as is.
